File: Java/Lemonade Change.java
public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public boolean lemonadeChange(int[] bills) {
        int five = 0, ten = 0;

        for (int bill : bills) {
            if (bill == 5) {
                five += 1;
            } else if (bill == 10 && five > 0) {
                five -= 1;
                ten += 1;
            } else if (bill == 20 && ten > 0 && five > 0) {
                five -= 1;
                ten -= 1;
            } else if (bill == 20 && five >= 3) {
                five -= 3;
            } else {
                return false;
            }
        }

        return true;
    }
}

File: Python/Lemonade Change.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def lemonadeChange(self, bills: List[int]) -> bool:
        five = ten = 0

        for bill in bills:
            if bill == 5:
                five += 1
            elif bill == 10 and five > 0:
                ten += 1
                five -= 1
            elif bill == 20 and ten > 0 and five > 0:
                five -= 1
                ten -= 1
            elif bill == 20 and five >= 3:
                five -= 3
            else:
                return False

        return True
